src/blog/oilpan-pointer-compression.md

```diff
@@ -0,0 +1,247 @@
+| ------------ | ---------------------------------------: | ----------------------------------------: |
+| heap pointer | <tt>U<sub>31</sub>...U<sub>1</sub>1</tt> | <tt>L<sub>31</sub>...L<sub>2</sub>00</tt> |
+| `nullptr`    | <tt>0...0</tt>                           | <tt>0...000</tt>                          |
+| sentinel     | <tt>0...0</tt>                           | <tt>0...010</tt>                          |
+<!-- markdownlint-enable no-inline-html -->
+:::
+
+Compression generates a compressed value by merely right-shifting by one and truncating away the upper half of the value.  In this way, the alignment bit (which now becomes the most significant bit of the compressed value) signals a valid heap pointer.
+
+:::table-wrapper
+| C++                                             | x64 assembly  |
+| :---------------------------------------------- | :------------ |
+| ```cpp                                          | ```asm        \
+| uint32_t Compress(void* ptr) {                  | mov rax, rdi  \
+|   return ((uintptr_t)ptr) >> 1;                 | shr rax       \
+| }                                               | ```           \
+| ```                                             |               |
+:::
+
+The encoding for compressed values is thus as follows:
+
+:::table-wrapper
+<!-- markdownlint-disable no-inline-html -->
+|              | compressed value                          |
+| ------------ | ----------------------------------------: |
+| heap pointer | <tt>1L<sub>31</sub>...L<sub>2</sub>0</tt> |
+| `nullptr`    | <tt>0...00</tt>                           |
+| sentinel     | <tt>0...01</tt>                           |
+<!-- markdownlint-enable no-inline-html -->
+:::
+
+Note that this allows for figuring out whether a compressed value represents a heap pointer, `nullptr`, or the sentinel value, which is important to avoid useless decompressions in user code (see below).
+
+The idea for decompression then is to rely on a specifically crafted base pointer, in which the least significant 32 bits are set to 1.
+
+:::table-wrapper
+<!-- markdownlint-disable no-inline-html -->
+|              | upper half                               | lower half     |
+| ------------ | ---------------------------------------: | -------------: |
+| base         | <tt>U<sub>31</sub>...U<sub>1</sub>1</tt> | <tt>1...1</tt> |
+<!-- markdownlint-enable no-inline-html -->
+:::
+
+
+The decompression operation first sign extends the compressed value and then left-shifts to undo the compression operation for the sign bit. The resulting intermediate value is encoded as follows
+
+:::table-wrapper
+<!-- markdownlint-disable no-inline-html -->
+|              | upper half     | lower half                                |
+| ------------ | -------------: | ----------------------------------------: |
+| heap pointer | <tt>1...1</tt> | <tt>L<sub>31</sub>...L<sub>2</sub>00</tt> |
```

Comment:
Please change
"tt>L<sub>31</sub>...L<sub>2</sub>00</tt>"
to
"tt>L<sub>31</sub>...L<sub>3</sub>000</tt>"